### PR TITLE
Add config_settings support for build backend calls

### PR DESCRIPTION
**Pull request by nschloe**
This is a first (non-working) stab at https://github.com/tox-dev/tox/issues/3080. I tried to add a config setting just like https://github.com/tox-dev/tox/pull/1891, but it doesn't seem to get passed on to `self.core`. It's supposed to be used like

```ini
[testenv]
config_settings = 
  abc = xyz
```

Any hints on what's missing?

### CODE REVIEW

File: src/tox/tox_env/python/virtual_env/package/pyproject.py [hunk 1]
```diff
@@ -214,7 +214,10 @@ def perform_packaging(self, for_env: EnvConfigSet) -> list[Package]:
         elif of_type == "sdist":
             self.setup()
             with self._pkg_lock:
-                sdist = self._frontend.build_sdist(sdist_directory=self.pkg_dir).sdist
+                print("cc", self.core["config_settings"])
```
Comment (gaborbernat) [inline src/tox/tox_env/python/virtual_env/package/pyproject.py:217]: I don't think this is part of the core settings, and I don't think it should be. This is the packaging environment setting in my view.

Comment (nschloe) [reply src/tox/tox_env/python/virtual_env/package/pyproject.py:217]: I'm not really sure what this means. What part of the code is the _packaging env setting_?

Comment (nschloe) [reply src/tox/tox_env/python/virtual_env/package/pyproject.py]: Got it.

**Comment by gaborbernat** [comment]
You likely want to have a base settings and then allow overriding for wheel/sdist/editable.

**Comment by nschloe** [comment]
Do you mean something like

```ini
[testenv:.pkg]
pep517_config_settings_base = 
  a = b
  x = yz
pep517_config_settings_sdist = 
  a = c
  x = z
pep517_config_settings_wheel = 
  a = W
  x = W
```
? In the interest of keeping things simple, I'm not convinced that's necessary. After all, there's a separate setting for sdist/wheel/editable builds, and when the user wants to change that, she has to touch the config _anyway_. E.g., from

```ini
[testenv:.pkg]
package = editable
pep517_config_settings = 
  a = b
  x = yz
```

to

```ini
[testenv:.pkg]
;package = editable
;pep517_config_settings = 
;  a = b
;  x = yz
package = sdist
pep517_config_settings = 
  a = S
  x = S
```
What do you think?

**Comment by gaborbernat** [comment]
Multiple different tox environments can reuse the same packaging environment, so it's important they are configurred differently:

```ini
[testenv]
wheel_build_env = .pkg

[testenv:wheel]
package = wheel


[testenv:sdist]
package = sdist

[testenv:editable]
package = editable
```

All these 3 envs will use the same `.pkg` and its settings even though they build differently.

**Comment by nschloe** [comment]
Ah, so what should the config look like exactly?

**Comment by gaborbernat** [comment]
```
[testenv]
wheel_build_env = .pkg

[testenv:wheel]
package = wheel


[testenv:sdist]
package = sdist

[testenv:editable]
package = editable

[testenv:.pkg]
config_settings = 
  a = S
wheel_config_settings = 
  a = S
sdist_config_settings = 
  a = S
editable_config_settings = 
  a = S
```

If `wheel_config_settings` is not set defaults to `config_settings` and so on.

**Comment by nschloe** [comment]
If
```
config_settings = 
  a = S
  b = T
config_settings_wheel = 
  a = U
```
should the complete list of config settings by
```
  a = U
  b = T
```
or
```
  a = U
```
?

**Comment by gaborbernat** [comment]
Don't merge. Instead override.

File: src/tox/tox_env/python/virtual_env/package/pyproject.py [hunk 2]
```diff
@@ -424,3 +467,10 @@ def _wheel_directory(self) -> Iterator[Path]:
 @impl
 def tox_register_tox_env(register: ToxEnvRegister) -> None:
     register.add_package_env(Pep517VirtualEnvPackager)
+
+
+def merge_dicts(*dicts):
```
Comment (gaborbernat) [inline src/tox/tox_env/python/virtual_env/package/pyproject.py:472]: Don't merge dicts.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Add tests too please.

**Comment by nschloe** [comment]
> What about the `self._wheel_config_settings`?

`self._wheel_config_settings` is now overridden my either of the other `config_settings` (if present).

**Comment by nschloe** [comment]
> Add tests too please.

Where would they go? `tests/conftest.py`?

**Comment by nschloe** [comment]
I'm having substantial problems setting up a test. There seems to be no similar test prototype I could take inspiration from. For example `meta_dir` and `pkg_dir` don't appear in the tests at all. That's not mentioning that I have no idea how to test if the arguments are actually passed through.

**Comment by nschloe** [comment]
@gaborbernat I'm gonna need some guidance here.

**Comment by gaborbernat** [comment]
Will need https://github.com/tox-dev/pyproject-api/pull/99 first released.

File: docs/changelog/3090.feature.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add support for setting build backend ``config_settings`` in the configuration file - by :user:`gaborbernat`.
```
Comment (webknjaz) [inline docs/changelog/3090.feature.rst:1]: @nschloe did you forget to use your own username here?

Comment (nschloe) [reply docs/changelog/3090.feature.rst:1]: Not really. I made _some_ contributions here, but Gabor finished it all up.

File: docs/config.rst
Comment (webknjaz) [inline docs/config.rst:1]: @nschloe I don't see any mentions of just `config_settings` that is not hook-specific but shared. Did that end up not being implemented?

Comment (gaborbernat) [reply docs/config.rst:1]: We did not add that, but you can do that with config reference in your tox file.

Comment (webknjaz) [reply docs/config.rst:1]: @gaborbernat do you mean compute dynamically in a `toxfile.py`?

Comment (gaborbernat) [reply docs/config.rst:1]: No, I meant just set it once in `tox.ini` and use config reference to set it for every hook you care about.

Comment (webknjaz) [reply docs/config.rst:1]: @gaborbernat when you say “config reference”, what exactly do you mean? I'm guessing that's what I don't understand. Could you help me bridge the gap in understanding?

Comment (gaborbernat) [reply docs/config.rst:1]: https://tox.wiki/en/4.21.2/config.html#configuration-reference

Comment (gaborbernat) [reply docs/config.rst:1]: refering to another config in the config file

Comment (webknjaz) [reply docs/config.rst:1]: Ah, I didn't know about that new syntax. The old one was substitutions. Thanks!

**Pull request merged**